Iter5

k1nky opened the pull request.

k1nky committed:
init iter1

k1nky committed:
add tests for metric package

k1nky committed:
add tests for storage package

k1nky committed:
refactor server

k1nky committed:
init agent

k1nky committed:
fix statictest warnings

k1nky committed:
add apiclient package

k1nky committed:
add Makefile

k1nky committed:
fix staticcheck warnings

k1nky committed:
refactor

k1nky committed:
enable coverage analysis

k1nky committed:
add handlers package

k1nky committed:
use resty as a http client

k1nky committed:
use gin as a http server

k1nky committed:
add valueHandler

k1nky committed:
add allMetricsHandler

k1nky committed:
refactor: comment

k1nky committed:
init iter4

k1nky committed:
set flags for server

k1nky committed:
set flags for agent

k1nky committed:
add tests for agent package

k1nky committed:
init iter5

k1nky committed:
set server config options from environment variables

k1nky committed:
update Makefile

k1nky committed:
set agent config options from environment variables

k1nky committed:
update server tests

k1nky committed:
refactor

connorholt commented on cmd/server/main.go:
```diff
@@ -1,3 +1,31 @@
+)
+
+var (
+	config *Config
```
лучше минимизировать кол-во глобальных переменных, тем более эта переменная только в этом пакет, давай сделаем отдельно пакет config куда уберем структуру и в main() можно будет делать как то так:

```
cfg := config.New()
cfg.Parse()
```

connorholt commented on cmd/server/main.go:
```diff
@@ -1,3 +1,31 @@
+
+func main() {
+	var err error
+	if config, err = Parse(nil); err != nil {
```
Parse(nil) рекомендую избегать использования функций таким образом, go way это про простоту и очевидность, если приходится передавать в функцию nil, значит что-то не так в проектировании. Со стороны выглядит не понятно что за аргумент, приходится лезть внутрь и искать что это. Поэтому тут надо это убрать

connorholt commented on cmd/server/main.go:
```diff
@@ -1,3 +1,31 @@
+}
+
+func run() error {
+	srv := server.New(server.WithStorage(storage.NewMemStorage()))
```
With это паттерн про опции, все-таки storage это не опция, а внутри в сервер все равно есть проверка и потом выполняется storage.NewMemStorage(), тут можно убрать опции т.к они делают код гибкий но скрывают зависимость.
Если передать storage как прямую зависимость код будет предсказуемей

connorholt commented on internal/server/server.go:
```diff
@@ -0,0 +1,70 @@
+// Пакет server реализует сервер сбора метрик
+package server
```
Название пакета путает и сбивает с толку, мне было не просто понять почему server зависит от storage и почему он передается внутрь handlerа, 
но когда я разобрался понял что это не сервер а скорее основной функционал приложения, поэтому предлагаю переименовать пакет и внутри структуры

connorholt commented on internal/server/server.go:
```diff
@@ -0,0 +1,70 @@
+type Option func(*Server)
+
+// WithStorage задает опцию, определяющее какое хранилище использовать для метрик
+func WithStorage(storage storage.Storage) Option {
```
опции этому пакету не нужны, преждевременно их делать не стоит

connorholt commented on internal/server/server.go:
```diff
@@ -0,0 +1,70 @@
+
+// Server сбора метрик
+type Server struct {
+	storage storage.Storage
```
В go принято не использовать интерфейсы из других пакетов чтобы не делать сильную связанность, но можно в каждом пакете держать по интерфейсу, поэтому интерфейс из storage будет лежать (приватным) и тут и в пакете storage

connorholt commented on cmd/server/main.go:
```diff
@@ -1,3 +1,31 @@
+func run() error {
+	srv := server.New(server.WithStorage(storage.NewMemStorage()))
+	log.Println("server starting on ", config.Address)
+	handler := handler.New(srv)
```
я бы разделил handlers и router, чтобы сделать еще прозрачнее и было видно что роутер можно собрать из разных хендлеров а не из самого себя

connorholt commented on internal/agent/agent.go:
```diff
@@ -0,0 +1,178 @@
+
+// Option опция конфигурации агента.
+// Используются при создании нового агента через функцию New.
+type Option func(*Agent)
```
пакет опции нужен для настройки, например удобно настраивать подключение для базы и тд, но при подключении к базе мы не передаем хост и порт как опцию, также как и не отдаем базе данных самим определять дефолтное поведение, в агенте не нужны опции.

Сюда можно передать конфиг, с прямой зависимостью или прямо конкретные аргументы как интервал и адрес, но делая через получается что это не обязательные параметры и у пакета появляется дополнительная ответственность сделать дефолтные настройки, хотя это не его обязанность.

connorholt commented on internal/agent/agent.go:
```diff
@@ -0,0 +1,178 @@
+)
+
+type Agent struct {
+	storage storage.Storage
```
тут тоже самое про интерфейсы, как я уже писал, причем достаточно будет сделать тут интерфейс с парой методов а не со всем что есть в Storage

connorholt commented on internal/agent/agent.go:
```diff
@@ -0,0 +1,178 @@
+	for _, opt := range options {
+		opt(a)
+	}
+	if a.storage == nil {
```
все зависимости передаем явно

connorholt commented on internal/agent/agent.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+func (a Agent) report() error {
+	for _, name := range a.storage.GetNames() {
```
минор, выше есть список метрик, можно обойти их и тогда не нужно будет делать специфичный для хранилища метод GetNames

connorholt commented on internal/apiclient/apiclient.go:
```diff
@@ -0,0 +1,67 @@
+	"github.com/k1nky/ypmetrics/internal/metric"
+)
+
+const DefEndpointURL = "http://localhost:8080"
```
Не стоит делать дефолтный url в клиенте, это не его обязанность, это можно сделать в одном месте в коде в конфиге, а сейчас получается это логика перетекает из пакета в пакет

connorholt commented on internal/apiclient/apiclient.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+// UpdateMetric отправляет значение указанной метрики на сервер
+func (c *Client) UpdateMetric(metric metric.Measure) (err error) {
```
как мне кажется клиент не должен завязываться на метрики не на уровне названия функции не на уровне реализации, он должен делать простую обязанность по урлу отправить данные

```
client := client.New()
client.Send(metricUrl, value)
```

как то так

connorholt commented on internal/handler/handler.go:
```diff
@@ -0,0 +1,74 @@
+// Пакет handler реализует обработчик HTTP запросов к серверу сбора метрик
+package handler
```
про handler и route было выше, нужно разделить логику и оставить каждому пакету по своей отвественности. в реальном приложении хендлеров может быть много и они потом участвует в составлении роутера, но это они сами не есть роутер

connorholt commented on internal/metric/metric.go:
```diff
@@ -0,0 +1,190 @@
+	// StringValue возвращает значение метрики как строку
+	StringValue() string
+	// Update обновляет значение метрики
+	Update(value interface{}) error
```
метод возвращает error которая по коду не обрабатывается в большинстве случаев и только в одном месте проверяется err

connorholt commented on internal/metric/metric.go:
```diff
@@ -0,0 +1,190 @@
+
+// New создает новую метрики с указаным типом type, именем name и значением value.
+// Функция аналогична New за исключением, того что иницилизирует ее указанным значением.
+func NewWtihValue(typ Type, name string, value interface{}) (Measure, error) {
```
тут скорее всего опечатка в названии, и в комментарии

connorholt commented on internal/metric/metric.go:
```diff
@@ -0,0 +1,190 @@
+
+// New создает новую метрики с указаным типом type и именем name.
+// Ошибка будет != nil, если указан неверный тип или имя.
+func New(typ Type, name string) (Measure, error) {
```
возвращать ошибку в New в функциях-кострукторах не надо, это не go way, у тебя эта функция это фабрика поэтому если фабрика не знает как сделать структуру пусть возвращает nil.

Это функцию стоит упросить до одной обязанности и убрать излишний код

```
func New(metricType Type, name string) Measure {
	switch metricType {
	case TypeGauge:
		return &Gauge{
			Name: name,
		}
	case TypeCounter:
		return &Counter{
			Name: name,
		}
	}

	return nil
}
```

и если использовать потом как то так, будет очевиднее

```
factory.NewMetric()
```

connorholt commented on cmd/agent/flags.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func parseFromEnv() (*Config, error) {
+	type cfg struct {
```
выше уже есть структура struct, тут это не надо, можно упростить

connorholt commented on internal/metric/metric.go:
```diff
@@ -0,0 +1,190 @@
+	default:
+		return ErrInvalidValue
+	}
+	g.Value = v
```
обе структуры почти индентичны между собой, и получается столько дополнительного кода хотя они отличаются только g.Value = v и c.Value += v, это можно упростить

можно сделать одну структуру которая по переданному типу возьмет функцию подсчета  и тогда большую часть кода не надо будет дублировать

connorholt commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,70 @@
+
+// MemStorage хранилище метрик в памяти
+type MemStorage struct {
+	values map[string]metric.Measure
```
надо уйти от зависимости с метриками, это можно сделать если прямо тут описать интерфейс

connorholt commented on internal/storage/storage.go:
```diff
@@ -0,0 +1,70 @@
+
+func (ms *MemStorage) UpSet(metric metric.Measure) {
+	if m := ms.Get(metric.GetName()); m != nil {
+		m.Update(metric.GetValue())
```
выглядит запутанно, положить в storage метрику чтобы она внутри обновилась, в storage должны сетится значения и storage не должен вызывать Update

k1nky committed:
refactor after review

add package config

remove package server

add metric set

avoid using interface{} in metric package

k1nky commented:
- Добавил пакет config для конфигов сервера и агента;
- Убрал не нужные опции;
- Убрал пакет server, в текущей реализации пока не вижу в нем смысла. Для реализации сервера достаточно обработчкиков запросов (handler) и набора метрик (metric.Set);
- Разделил роутер и обработчики;
- Лишние интерфейсы убрал, оставил только интерфейс хранилища;
- Резделил метрики counter и gauge явно по типу, чтоб избежать излишнего использования interface{}.

connorholt commented on internal/metric/metric.go:
```diff
@@ -0,0 +1,63 @@
+}
+
+// String возвращает строковое предствление метрики в формате <тип><имя><значение>
+func (g Gauge) String() string {
```
хорошей практикой делать ресиверы одинаковые для всех функций, поэтому если есть методы где ресивер *Gauge, то лучше везде также использовать. 
_Struct Gauge has methods on both value and pointer receivers. Such usage is not recommended by the Go Documentation._

k1nky committed:
use the same reciever
